File: app/src/main/java/com/techbeloved/hymnbook/utils/Decompress.java
package com.techbeloved.hymnbook.utils;


import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

/**
 * @author jon
 * <p>
 * Usage
 * String zipFile = Environment.getExternalStorageDirectory() + "/the_raven.zip"; //your zip file location
 * String unzipLocation = Environment.getExternalStorageDirectory() + "/unzippedtestNew/"; // destination folder location
 * DecompressFast df= new DecompressFast(zipFile, unzipLocation);
 * df.unzip();
 * <p>
 * Add permissions
 * <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
 * <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
 */
public class Decompress {
    private String _zipFile;
    private String _location;

    public Decompress(String zipFile, String location) {
        _zipFile = zipFile;
        _location = location;

        _dirChecker("");
    }

    public void unzip() throws Exception {

        FileInputStream fin = new FileInputStream(_zipFile);
        ZipInputStream zin = new ZipInputStream(fin);
        ZipEntry ze;
        while ((ze = zin.getNextEntry()) != null) {

            if (ze.isDirectory()) {
                _dirChecker(ze.getName());
            } else {
                FileOutputStream fout = new FileOutputStream(_location + "/" + ze.getName());
                BufferedOutputStream bufout = new BufferedOutputStream(fout);
                byte[] buffer = new byte[1024];
                int n;
                while ((n = zin.read(buffer)) != -1) {
                    bufout.write(buffer, 0, n);
                }

                zin.closeEntry();
                bufout.close();
                fout.close();
            }

        }
        zin.close();


        File file = new File(_zipFile);
        file.delete();
    }

    private void _dirChecker(String dir) {
        File f = new File(_location + dir);

        if (!f.isDirectory()) {
            f.mkdirs();
        }
    }
}

File: app/src/main/java/com/techbeloved/hymnbook/utils/Decompress.kt
package com.techbeloved.hymnbook.utils

import java.io.File
import java.io.FileOutputStream
import java.util.zip.ZipFile
import java.util.zip.ZipInputStream
import javax.inject.Inject

class Decompress @Inject constructor() {
    fun unzip(zipFilePath: String, destinationDirPath: String) {
        val zipFile = ZipFile(zipFilePath)
        val entries = zipFile.entries()

        while (entries.hasMoreElements()) {
            val entry = entries.nextElement()

            val entryPath = File(destinationDirPath, entry.name)
            if (entry.isDirectory) {
                entryPath.mkdirs()
            } else {
                entryPath.createNewFile()

                val inputStream = zipFile.getInputStream(entry)
                val outputStream = FileOutputStream(entryPath)

                inputStream.copyTo(outputStream)

                inputStream.close()
                outputStream.close()
            }
        }

        zipFile.close()
    }

    fun unzip(zipInputStream: ZipInputStream, destinationDirPath: File) {
        zipInputStream.use { inputStream ->
            var entry = inputStream.nextEntry
            while (entry != null) {
                val entryPath = File(destinationDirPath, entry.name)
                if (entry.isDirectory) {
                    entryPath.mkdirs()
                } else {
                    if (entryPath.exists()) {
                        entryPath.delete()
                    }
                    val fileOutputStream = FileOutputStream(entryPath)
                    inputStream.copyTo(fileOutputStream)
                    fileOutputStream.close()
                }
                entry = inputStream.nextEntry
            }
        }
    }
}
